### DISCUSSION TITLE
Can an XHR POST action() request send back loader() data instead of making a subsequent GET request?

### DISCUSSION THREAD

**Discussion by daskalou** [remix-run/remix#4405] [Q&A]
I was watching the Remix Singles videos presented by Ryan Florence on Remix's YouTube channel:

https://www.youtube.com/playlist?list=PLXoynULbYuEDG2wBFSZ66b85EIspy3fy6

They were great for learning Remix better.

In the videos Ryan shows how Remix falls back to normal form posting if JavaScript is disabled (an awesome feature of Remix).

What I did notice though is that the fallback form posting felt faster than sending the form via fetch() / XHR.

I would have thought that the fetch() / XHR would have been faster, but then looking at the Network tab of Dev Tools, it looks like Remix made 2 requests:

(1) POST - Send form data
(2) GET - After (1) completes, retrieve updated loader() data

**My question is, is it possible to configure Remix's fetcher.Form to tell the server to call the loader() after the action() has completed, and include the loader()'s data in the same POST response?**

That way another roundtrip GET request is not needed, and thus would make the Remix UI feel snappier (assume no optimistic UI is used).

**Comment by akomm** [comment]
An action might trigger reload of multiple loaders. Not just the one in the same route. The result of action and loader are also separate, hence you have two hooks `useActionData` and `useLoaderData`. You can't merge the results reliably for all use-cases. 

You might use this:
https://github.com/remix-run/remix/discussions/4828

And then:
```ts
const actionUrl = useActionUrl()
// ... later useCallback/effect, etc.
xhr.open(actionUrl)
// ...
xhr.send(/* ... */)
```

It should also not implicitly call loader when you send to the generated URL. So you are free then to return the data in the action and use it. I don't think thought that in general that is a good idea if you have the data from loader also return it from action. You will have to handle the merge. This is the point were you notice why they are separate. Note: a response in loader and action can be anything independently. Its just a Response object you create. It doesn't have to be data. You might have a redirect in action, then the load doesn't need to be called at all.
